sled-storage/src/manager.rs

```diff
@@ -398,35 +400,75 @@ impl StorageManager {
         );
         self.state = StorageManagerState::Normal;
 
+        // Do we need to inform our consumers about any new disks?
         let mut send_updates = false;
 
-        // Disks that should be requeued.
-        let queued = self.queued_u2_drives.clone();
-        let mut to_dequeue = HashSet::new();
-        for disk in queued.iter() {
-            if self.state == StorageManagerState::QueueingDisks {
-                // We hit a transient error in a prior iteration.
-                break;
-            } else {
-                match self.add_u2_disk(disk.clone()).await {
-                    Err(_) => {
+        let mut futs = FuturesUnordered::new();
+
+        // Create new disks concurrently
+        for raw_disk in self.queued_u2_drives.iter().cloned() {
+            let key_requester = self.key_requester.clone();
+            let log = self.log.clone();
+            let raw_disk2 = raw_disk.clone();
+            futs.push(async move {
+                (
+                    raw_disk,
+                    Disk::new(&log, raw_disk2, Some(&key_requester)).await,
+                )
+            })
+        }
+
+        // Check to see if any of our disk creations succeeded.
+        // If so, insert them into `self.resources`.
+        while let Some((raw_disk, res)) = futs.next().await {
+            match res {
+                Ok(disk) => match self.resources.insert_disk(disk) {
+                    Err(err) => {
                         // This is an unrecoverable error, so we don't queue the
                         // disk again.
-                        to_dequeue.insert(disk);
+                        error!(
+                            self.log,
+                            "Persistent error: {err}: not queueing disk";
+                            "disk_id" => ?raw_disk.identity()
+                        );
+                        self.queued_u2_drives.remove(&raw_disk);
                     }
                     Ok(AddDiskResult::DiskInserted) => {
                         send_updates = true;
-                        to_dequeue.insert(disk);
+                        self.queued_u2_drives.remove(&raw_disk);
                     }
                     Ok(AddDiskResult::DiskAlreadyInserted) => {
-                        to_dequeue.insert(disk);
+                        self.queued_u2_drives.remove(&raw_disk);
                     }
-                    Ok(AddDiskResult::DiskQueued) => (),
+                    Ok(AddDiskResult::DiskQueued) => {
+                        error!(
```

Comment:
Should we panic here? If for some reason this error code came back, we probably should set `self.state = StorageManagerState::QueueingDisks`

Reply:
I don't really like the idea of panicking here, but I'm not sure I understand all these pathways.

This codepath looks pretty redundant with the call to `add_u2_disk`, which calls `Disk::new` and also sets the `QueueingDisks` state when we match on an underlying error from `self.resources.insert_disk`.

Is there a reason we aren't using that method anymore? I'm kinda nervous by so much state-setting and checking from different possible pathways. Specifically: If I were to edit this codebase, I'm really unclear "who is responsible for setting `self.state`, caller, callee, or my current function".